### PR TITLE
Universal build tasks

### PR DESCRIPTION
**Pull request by shubham90**

### CODE REVIEW

File: Tasks/Common/nuget-task-common/LocationHelpers.ts [hunk 1]
```diff
@@ -148,7 +158,7 @@ export function getConnectionDataForArea(
 }
 
 export function getNuGetConnectionData(serviceUri: string, accessToken: string): Q.Promise<locationApi.ConnectionData> {
-    return getConnectionDataForArea(serviceUri, "nuget", "b3be7473-68ea-4a81-bfc7-9530baaa19ad", accessToken);
+    return getConnectionDataForArea(serviceUri, "nuget", "b3be7473-68ea-4a81-bfc7-9530baaa19ad", "fakeUsername", accessToken);
```
Comment (zarenner) [inline Tasks/Common/nuget-task-common/LocationHelpers.ts]: nit: usually we use something like "token" or "vsts" here

File: make-options.json [hunk 2]
```diff
@@ -99,6 +99,7 @@
         "SqlAzureDacpacDeploymentV1",
         "SqlDacpacDeploymentOnMachineGroupV0",
         "SshV0",
+        "UPackCommandV0",
```
Comment (zarenner) [inline make-options.json]: Rename to Universal?

File: Tasks/UPackCommandV0/upackpublish.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        tl.setResult(tl.TaskResult.Failed, tl.loc("PackagesFailedToPublish"));
+    }
+}
+
+function publishPackageUsingArtifactTool(publishDir: string, options: artifactToolRunner.IArtifactToolOptions) {
+    let command = new Array<string>();
+
+    command.push("upack", "publish",
```
Comment (zarenner) [inline Tasks/UPackCommandV0/upackpublish.ts]: universal

File: Tasks/Common/nuget-task-common/LocationHelpers.ts [hunk 4]
```diff
@@ -103,9 +114,9 @@ export function getConnectionDataForArea(
     serviceUri: string,
     areaName: string,
     areaId: string,
+    username: string,
     accessToken: string): Q.Promise<locationApi.ConnectionData> {
```
Comment (zarenner) [inline Tasks/Common/nuget-task-common/LocationHelpers.ts]: consider rename: really 'password', which happens to be an access token

File: Tasks/UPackCommandV0/Common/ArtifactToolRunner.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    accountUrl: string;
+    packageName: string;
+    packageVersion: string;
+    authInfo: auth.UPackExtendedAuthInfo;
+}
+
+function getOptions(): IExecOptions{
+    let result: IExecOptions = <IExecOptions>{
+        cwd: process.cwd(),
+        env: process.env,
```
Comment (zarenner) [inline Tasks/UPackCommandV0/Common/ArtifactToolRunner.ts]: Can you clone the environment and inject the PAT only into the artifacttool process environment, instead of the current process environment?

File: Tasks/UPackCommandV0/Common/ArtifactToolRunner.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import stream = require("stream");
+import {IExecOptions, IExecSyncResult} from "vsts-task-lib/toolrunner";
+
+export interface IArtifactToolOptions {
+    artifactToolPath: string;
+    feedId: string;
+    accountUrl: string;
+    packageName: string;
+    packageVersion: string;
+    authInfo: auth.UPackExtendedAuthInfo;
```
Comment (zarenner) [inline Tasks/UPackCommandV0/Common/ArtifactToolRunner.ts]: Where is authInfo used?

Comment (shubham90) [reply Tasks/UPackCommandV0/Common/ArtifactToolRunner.ts]: It's not used now. I can remove it. I put it there initially to support username/password.

File: Tasks/Common/nuget-task-common/LocationHelpers.ts (hunk 4 above)
Comment (shubham90) [reply Tasks/Common/nuget-task-common/LocationHelpers.ts]: cool, will do.

File: Tasks/UPackCommandV0/Common/ArtifactToolRunner.ts (hunk 5 above)
Comment (shubham90) [reply Tasks/UPackCommandV0/Common/ArtifactToolRunner.ts]: The artifact tool process env and current process env is same right now. Is there any benefit of creating a clone?

File: make-options.json (hunk 2 above)
Comment (mitchdenny) [reply make-options.json]: Yep.

Comment (shubham90) [reply make-options.json]: updated

File: Tasks/UniversalPackagesV0/Common/ArtifactToolRunner.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    feedId: string;
+    accountUrl: string;
+    packageName: string;
+    packageVersion: string;
+}
+
+function getOptions(): IExecOptions{
+    let result: IExecOptions = <IExecOptions>{
+        cwd: process.cwd(),
+        env: new Object(process.env),
```
Comment (zarenner) [inline Tasks/UniversalPackagesV0/Common/ArtifactToolRunner.ts]: new Object doesn't actually do a copy

File: Tasks/UniversalPackagesV0/Strings/resources.resjson/en-US/resources.resjson [hunk 8]
```diff
@@ -0,0 +1,78 @@
+{
```
Comment (zarenner) [inline Tasks/UniversalPackagesV0/Strings/resources.resjson/en-US/resources.resjson]: Throughout: I think we want to prefer "Universal Packages" over "universal packages" or "Universal packages". 
@alexmullans @mitchdenny is that correct?

File: Tasks/UniversalPackagesV0/Common/ArtifactToolUtilities.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+        artifactToolPath = await toollib.cacheDir(unzippedToolsDir, "ArtifactTool", artifactToolUri.version);
+    }
+    else{
+        tl.debug(tl.loc("Info_ResolvedToolFromCache", artifactToolPath));
+    }
+    return getArtifactToolLocation(artifactToolPath);
+}
+
+// set the console code page to "UTF-8"
+export function setConsoleCodePage() {
```
Comment (zarenner) [inline Tasks/UniversalPackagesV0/Common/ArtifactToolUtilities.ts]: Why is this necessary / is this something every task does?

Comment (shubham90) [reply Tasks/UniversalPackagesV0/Common/ArtifactToolUtilities.ts]: our nuget and dotnet task does it, I don't think it's necessary though. Copy pasted as a boiler plate.

Comment (zarenner) [reply Tasks/UniversalPackagesV0/Common/ArtifactToolUtilities.ts]: I would confirm it's needed / avoid bloat if possible.

Comment (jmyersmsft) [reply Tasks/UniversalPackagesV0/Common/ArtifactToolUtilities.ts]: It was necessary at one point. They may have changed the build agent to do this though

**Comment by alexmullans** [comment]
One more note: for "Path to file(s) to publish", the default _should_ be $(Build.ArtifactStagingDirectory), as we'll encourage users to copy their files there before publishing.

File: Tasks/UniversalPackagesV0/universaldownload.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+export async function run(artifactToolPath: string): Promise<void> {
+    let buildIdentityDisplayName: string = null;
+    let buildIdentityAccount: string = null;
+    try {
+        // Get directory to publish
+        let downloadDir: string = tl.getInput("downloadDirectory");
+        if (downloadDir.length < 1)
+        {
+            //todo
+            tl.warning(tl.loc("Info_NoPackagesMatchedTheSearchPattern"));
```
Comment (zarenner) [inline Tasks/UniversalPackagesV0/universaldownload.ts]: ???

Comment (zarenner) [reply Tasks/UniversalPackagesV0/universaldownload.ts]: Oh, the resource string is just poorly named, the message "No directory matched the search pattern." is somewhat reasonable but could be better. there's a todo there though?

Comment (shubham90) [reply Tasks/UniversalPackagesV0/universaldownload.ts]: yeah, I just missed my todo. Sorry about that and thanks for pointing this.

**Comment by shubham90** [comment]
@alexmullans Having feed name in yaml file is not possible because we need the feed id for populating the packages drop down. There is no feed api that takes feed name. And AFAIK, we can assign only one value to a json object.
